Improve df to tables writer

mmkekic opened the pull request:
This PR removes underscore in front of store_pandas_as_tables writer since the method is being used outside the module. It also modifies the method to improve speed when writing large dataframes, no tests needed because it is just internal change.

mmkekic committed:
Speedup writer by appending full numpy record array.

The writing is done by appending a numpy record array to the
pytable. It also adds checks of whether the type conversion is possible.

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -58,9 +58,12 @@ def _store_pandas_as_tables(h5out : tb.file.File, df : pd.DataFrame, group_name
+    arr = df.to_records(index=False)
+    #hack to transform object numpy dtype to string
+    data_types = [(col, arr[col].dtype if arr[col].dtype != 'O' else \
+                   f'S{str_col_length}') for col in arr.dtype.names]
```
It's a bit hard to read. Can we define a function (even inline) to do that?

mmkekic replied:
How about now?

gonzaponte reviewed (commented):
I think the tests are also checking that the types remain unchanged after writing & reading, right? I would like to have a specific one, but it's ok as is.

mmkekic committed:
Add test raise  warning and error.

Add tests that check the warining is raised if the string is too long,
and a test that check TableMismatch error is raised if the numeric
types are different.

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -29,25 +29,28 @@ def load_dsts(dst_list, group, node):
             tabledef[colname] = tb.StringCol(str_col_length, pos=indx)
         else:
             tabledef[colname] = tb.Col.from_type(coltype, pos=indx)
     return tabledef
```
Actually I don't think `_make_tabledef` cares about the length of the object. I would revert this back to the original version and perform the check in `store_pandas_as_tables`. What do you think?

mmkekic replied:
Agree. I added a check in `store_pandas_as_tables ` directly. I also modified to cast numpy records array to the table type if it was an already existing one. However I feel like this 3x search for strings and type castings can be somehow better optimized...

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -58,9 +61,16 @@ def _store_pandas_as_tables(h5out : tb.file.File, df : pd.DataFrame, group_name
+
+    arr = df.to_records(index=False)
+    #hack to transform object numpy dtype to string
+    def _cast_type(dtype : np.dtype):
```
I understand where the underscore comes from, but since it is inside a function it doesn't need to have special syntax :)

gonzaponte commented on invisible_cities/io/dst_io_test.py:
```diff
@@ -155,4 +156,14 @@ def test_store_pandas_as_tables_raises_warning_empty_dataframe(config_tmpdir, em
+    table_name = 'table_name_3'
+    with tb.open_file(filename, 'w') as h5out:
+        with pytest.warns(UserWarning, match='dataframe contains strings longer than allowed'):
+            store_pandas_as_tables(h5out, df, group_name, table_name)
```
I suggest that here you set explicitly the string column length in the call to `store_pandas_as_tables` to make the test more obvious.

gonzaponte commented on invisible_cities/io/dst_io_test.py:
```diff
@@ -118,7 +118,7 @@ def test_store_pandas_as_tables_2df(config_tmpdir, df1, df2):
         store_pandas_as_tables(h5out, df2, group_name, table_name)
     df_read = load_dst(filename, group_name, table_name)
-    assert_dataframes_close(df_read, pd.concat([df1, df2]).reset_index(drop=True), False, rtol=1e-5)
+    assert_dataframes_equal(df_read, pd.concat([df1, df2]).reset_index(drop=True))
```
these dataframes contain floats, `assert_dataframes_close` is more appropriate.

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -39,28 +39,43 @@ def _make_tabledef(column_types : pd.Series, str_col_length : int=32) -> dict:
+        #hack to transform object numpy dtype to string
+        def _cast_type(dtype : np.dtype):
+            if dtype != 'O': return dtype
+            else:            return f'S{str_col_length}'
```
Remove underscore.

Don't you find it weird to test for `!=` instead of `==`? There is nothing wrong with it, but it feels counterintuitive.

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -39,28 +39,43 @@ def _make_tabledef(column_types : pd.Series, str_col_length : int=32) -> dict:
+                warnings.warn(f'dataframe contains strings longer than allowed', UserWarning)
+        else:
+            if arr.dtype[colname] != data_types[colname]:
+                    warnings.warn(f'dataframe numeric types not consistent with the table existing ones', UserWarning)
```
I think this part belongs to the case in which there is already a table object stored. It doesn't make sense to check it for both cases

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -39,28 +39,43 @@ def _make_tabledef(column_types : pd.Series, str_col_length : int=32) -> dict:
+
+    #check which columns are of type object and throw warning if length exceeds
+    for colname in df.columns:
+        if (df[colname].dtype.name == 'object'):
```
Maybe this could be written as
```
for colname, col in filter(lambda (_, c): c.dtype.name == "object", df.items()):
    if col.str.len().max() > data_types[colname].itemsize:
        warnings.warn...
```
or
```
for colname, col in df.items():
    if col.dtype.name == "object" and col.str.len().max() > data_types[colname].itemsize:
        warnings.warn...
```

mmkekic replied:
Right... Let me try to rewrite the whole function in a more sensible way...

mmkekic committed:
Fix tests for modified _make_tabledef input

Also removes False for type checking in assert_dataframe_close

mmkekic committed:
Order arr columns before casting and saving

In case new dataframe columns do not have the same order of columns as
the already existing table

mmkekic committed:
Add test for store_pandas_as_tables with unordered dataframes

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -29,53 +29,52 @@ def load_dsts(dst_list, group, node):
+    if len(arr) == 0:
+        warnings.warn(f'dataframe is empty', UserWarning)
+    else:
+        _can_cast(arr, data_types)
```
I suggest a better name like `_check_castability` or something like that. The use of "can" suggests that it returns a boolean value.

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -75,6 +75,7 @@ def store_pandas_as_tables(h5out : tb.file.File, df : pd.DataFrame, group_name :
     else:
         _can_cast(arr, data_types)
-        arr = arr.astype(data_types)
+        columns = list(data_types.names)
+        arr = arr[columns].astype(data_types)
```
where is the sorting?

mmkekic replied:
It is not sorted, is just assured that the columns of record array have the same order as the ones in table

gonzaponte replied:
OK. Got slightly confused by the commit message.

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -29,38 +29,53 @@ def load_dsts(dst_list, group, node):
+            warnings.warn(f'dataframe numeric types not consistent with the table existing ones', UserWarning)
+
+
+def store_pandas_as_tables(h5out : tb.file.File, df : pd.DataFrame, group_name : str, table_name : str, compression : str='ZLIB4', descriptive_string : [str]="", str_col_length : int=32) -> None:
```
I know it was already like that, but can we take the opportunity to split this into many for readability?

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -29,38 +29,53 @@ def load_dsts(dst_list, group, node):
+            if max_str_length > table_types[name].itemsize:
+                warnings.warn(f'dataframe contains strings longer than allowed', UserWarning)
+        elif arr_types[name] != table_types[name]:
+            warnings.warn(f'dataframe numeric types not consistent with the table existing ones', UserWarning)
```
Shouldn't these be actual errors rather than just warnings?

mmkekic replied:
I am not sure, probably we do want error in case you try to store string in numeric column, but we do not want to stop the process if someone tries to put an Int32 value into Int64 columns... And writing all possible combinations that make sense seems like a bit tedious job, so I opted for 'leave it to user responsibility to check the warning'

gonzaponte replied:
Mmm, if you think about the cities, that shouldn't happen as the types should be consistent through the iterations. If you think about other usages, I don't see that it is a big deal for the user to deal with slightly different types.

For the first warning the code will probably just chop part of the string, and the second one would either crash or store useless data. And, being realistic, warnings are not usually checked..

gonzaponte commented on invisible_cities/io/dst_io.py:
```diff
@@ -29,38 +29,60 @@ def load_dsts(dst_list, group, node):
+            if max_str_length > table_types[name].itemsize:
+                warnings.warn(f'dataframe contains strings longer than allowed', UserWarning)
+        elif not np.can_cast(arr_types[name], table_types[name], casting='same_kind'):
+            raise TableMismatch(f'dataframe numeric types not consistent with the table existing ones')
```
```suggestion
def _check_castability(arr : np.ndarray, table_types : np.dtype):
    arr_types = arr.dtype
    
    if set(arr_types.names) != set(table_types.names):
        raise TableMismatch(f'dataframe differs from already existing table structure')
    
    for name in arr_types.names:
        if arr_types[name].name == 'object':
            max_str_length = max(map(len, arr[name]))
            if max_str_length > table_types[name].itemsize:
                warnings.warn(f'dataframe contains strings longer than allowed', UserWarning)
        
        elif not np.can_cast(arr_types[name], table_types[name], casting='same_kind'):
            raise TableMismatch(f'dataframe numeric types not consistent with the table existing ones')
```
Helps readability.

gonzaponte commented on invisible_cities/io/dst_io_test.py:
```diff
@@ -155,4 +155,36 @@ def test_store_pandas_as_tables_raises_warning_empty_dataframe(config_tmpdir, em
+
+
+@given(df=dataframe)
+def test_store_pandas_as_tables_raises_warning_inconsistent_types(config_tmpdir, df):
```
```suggestion
def test_store_pandas_as_tables_raises_TableMismatch_inconsistent_types(config_tmpdir, df):
```

gonzaponte reviewed (commented):
After these cosmetical changes we are good to go

gonzaponte reviewed (approved):
This PR makes the function `store_pandas_as_table` public, as it should be. It also improves the performance of the writer and adds type checks for robustness. A bunch of tests have been added, which is always good.

Good job!

mmkekic committed:
Remove _ in front of store_pandas_as_tables

The function is not used only in this module so it shouldnt be
private.
